### PR TITLE
Update tutorial

### PR DESCRIPTION
**Pull request by delta2323**
This PR updates the tutorial to follow the up-to-date usage of Chainer v2. I also update the examples in the tutorial that are obsolete.

### CODE REVIEW

File: docs/source/tutorial/convnet.rst [hunk 1]
```diff
@@ -107,11 +105,20 @@ computations in the model. Just prepare the data, then give it to the model.
 The way this works is the resulting output :class:`~chainer.Variable` from the
 forward computation has a :meth:`~chainer.Variable.backward` method to perform
 autograd. In the above model, :meth:`__call__` has a ``if`` statement at the
-end to switch its behavior by the model's running mode, i.e., training mode or
-not. When it's in training mode, this method returns the output value of the
+end to switch its behavior by the Chainer's running mode, i.e., training mode or
+not. Chainer presents the running mode as a global variable `chainer.config.train`.
+When it's in training mode, this method returns the output value of the
```
Comment (niboshi) [inline docs/source/tutorial/convnet.rst]: How about changing the pronoun "this method" to the explicit expression i.e. "`__call__`"?
It has become far away from the previous appearance.

File: docs/source/tutorial/convnet.rst [hunk 2]
```diff
@@ -107,11 +105,20 @@ computations in the model. Just prepare the data, then give it to the model.
...
 last layer as is to compute the loss later on, otherwise it returns a
 prediction result by calculating :meth:`~chainer.functions.softmax`.
 
+.. note::
+
+  In Chainer v1, if a function or link behaves differently in
+  training and other modes, it is common that it holds an attribute
+  that represents its running mode or is provided with the mode
+  form outside as an argument. In Chainer v2, it is recommended to use
```
Comment (niboshi) [inline docs/source/tutorial/convnet.rst]: v1 part should be in the past tense.

File: docs/source/tutorial/function.rst [hunk 3]
```diff
@@ -471,8 +471,7 @@ We can also initialize the parameter after the initialization by the :meth:`Link
    class EltwiseParamProduct(Link):
        def __init__(self, shape):
            super(EltwiseParamProduct, self).__init__()
-           self.add_param('W', shape)
-           self.W.data[...] = np.random.randn(*shape)
+           self.add_param('W', shape, initializers.Normal(scale=1.))
```
Comment (niboshi) [inline docs/source/tutorial/function.rst]: `add_param` is deprecated. It should be rewritten in the new method.

File: docs/source/tutorial/function.rst [hunk 4]
```diff
@@ -585,7 +585,8 @@ This is a test example of :func:`functions.relu` function
            y.grad = np.random.randn(3, 2).astype(np.float32)
            y.backward()
 
-           f = lambda: (F.relu(x).data,)
+           def f():
+               return F.relu(x).data,
```
Comment (niboshi) [inline docs/source/tutorial/function.rst]: Put a blank line after the function definition.

File: docs/source/tutorial/function.rst [hunk 5]
```diff
@@ -512,9 +511,10 @@ In order to make a convenient module, let's wrap it into a link:
 
    class Linear(Link):
        def __init__(self, in_size, out_size):
-           super(Linear, self).__init__(W=(out_size, in_size), b=out_size)
-           self.W.data[...] = np.random.randn(out_size, in_size) / math.sqrt(in_size)
-           self.b.data.fill(0)
+           super(Linear, self).__init__()
+           self.add_param('W', (out_size, in_size),
+                          initializers.Normal(1. / math.sqrt(in_size)))
+           self.add_param('b', (out_size,), initializers.Constant(0))
```
Comment (niboshi) [inline docs/source/tutorial/function.rst]: ditto (`add_param` is deprecated)

File: docs/source/tutorial/basic.rst [hunk 6]
```diff
@@ -178,6 +178,22 @@ An instance of the Linear link acts like a usual function:
    array([[ 3.1757617 ,  1.75755572],
           [ 8.61950684,  7.18090773]], dtype=float32)
 
+.. note::
+
+  Sometimes it is cumbersome to compute the dimension of input space.
```
Comment (niboshi) [inline docs/source/tutorial/basic.rst]: Put "the" before "input space".

File: docs/source/tutorial/basic.rst [hunk 7]
```diff
@@ -178,6 +178,22 @@ An instance of the Linear link acts like a usual function:
...
+  The linear link and some of (de)convolution links can omit the input dimension
+  in their instantiation and infer it from the first mini-batch.
+
+  For example, the following line creates a linear link whose output dimension
+  is two::
+
+      f = L.Linear(2)
+
+  If we feed a mini-batch of shape ``(N, M)``, the input dimension will be inferred as ``M``,
+  which means ``f.W`` is a 2 x M matrix.
```
Comment (niboshi) [inline docs/source/tutorial/basic.rst]: "is" -> "will be"

File: docs/source/tutorial/function.rst [hunk 8]
```diff
@@ -603,4 +604,32 @@ The next two lines compute numerical gradient using the same forward function wi
 And at last, we compare these two results elementwise.
 Note that the above test code can be easily modified to test GPU version just by replacing CPU arrays to GPU arrays.
 
+In most case, we do not write the code like the above explicitly because Chainer
```
Comment (niboshi) [inline docs/source/tutorial/function.rst]: "case" -> "cases"

File: docs/source/tutorial/function.rst [hunk 9]
```diff
@@ -603,4 +604,32 @@ The next two lines compute numerical gradient using the same forward function wi
...
+offers a utility function :func:`chainer.gradient_check.check_backward` that does
+follow this procedure.
```
Comment (niboshi) [inline docs/source/tutorial/function.rst]: How abount simply "that follows"?

**Comment by niboshi** [comment]
Please check this review as well.

> Put a blank line after the function definition.

File: docs/source/tutorial/basic.rst [hunk 10]
```diff
@@ -242,24 +258,34 @@ Then, what we have to do here is just define the above class as a subclass of Ch
...
-   Such links and chains are callable and behave like regular functions of Variables.
-
 It shows how a complex chain is constructed by simpler links.
 Links like ``l1`` and ``l2`` are called *child links* of MyChain.
 **Note that Chain itself inherits Link**.
 It means we can define more complex chains that hold MyChain objects as their child links.
 
+.. note::
+
+   We often define a single forward method of a link by ``__call__`` operator.
```
Comment (bkvogel) [inline docs/source/tutorial/basic.rst]: by the ``__call__`` operator.

File: docs/source/tutorial/function.rst (hunk 4 above)
Comment (delta2323) [inline docs/source/tutorial/function.rst]: fixed

**Pull request merged**